music_assistant/server/providers/deezer/helpers.py

```diff
@@ -0,0 +1,576 @@
+        artists.append(await parse_artist(artist=artist, mass=mass))
+    return artists
+
+
+async def search_and_parse_albums(
+    mass, client: deezer.Client, query: str, limit: int = 5
+) -> list[Album]:
+    """Search for album and parse them."""
+    deezer_albums = await search_album(client=client, query=query, limit=limit)
+    albums = []
+    for album in deezer_albums:
+        albums.append(await parse_album(album=album, mass=mass))
+    return albums
+
+
+async def search_and_parse_playlists(
+    mass, client: deezer.Client, query: str, limit: int = 5
+) -> list[Playlist]:
+    """Search for playlists and parse them."""
+    deezer_playlists = await search_playlist(client=client, query=query, limit=limit)
+    playlists = []
+    for playlist in deezer_playlists:
+        playlists.append(await parse_playlist(playlist=playlist, mass=mass))
+    return playlists
+
+
+def _md5(data, data_type="ascii"):
+    md5sum = hashlib.md5()
+    md5sum.update(data.encode(data_type))
+    return md5sum.hexdigest()
+
+
+def get_blowfish_key(track_id):
+    """Get blowfish key to decrypt a chunk of a track."""
+    secret = "g4el58wc" + "0zvf9na1"
+    id_md5 = _md5(track_id)
+    bf_key = ""
+    for i in range(16):
+        bf_key += chr(ord(id_md5[i]) ^ ord(id_md5[i + 16]) ^ ord(secret[i]))
+    return bf_key
+
+
+def decrypt_chunk(chunk, blowfish_key):
+    """Decrypt a given chunk using the blow fish key."""
+    cipher = Blowfish.new(
+        blowfish_key.encode("ascii"), Blowfish.MODE_CBC, b"\x00\x01\x02\x03\x04\x05\x06\x07"
+    )
+    return cipher.decrypt(chunk)
+
+
+def _gw_api_call(session, method, api_token="null", args=None, params=None):
+    if params is None:
+        params = {}
+    if args is None:
+        args = {}
+    p = {"api_version": "1.0", "api_token": api_token, "input": "3", "method": method}
+    p.update(params)
+    result_json = session.post(
+        "http://www.deezer.com/ajax/gw-light.php",
+        params=p,
+        timeout=30,
+        json=args,
+        headers={
+            "User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) "
+            "Chrome/79.0.3945.130 Safari/537.36"
+        },
+    ).json()
+    return result_json
+
+
+async def get_deezer_track_url(session, track_id):
+    """Get the URL for a given track id."""
+
+    def _get_deezer_track_url():
+        user_data = _gw_api_call(session, "deezer.getUserData")
+        api_token = user_data["results"]["checkForm"]
+        license_token = user_data["results"]["USER"]["OPTIONS"]["license_token"]
+        sng_data = _gw_api_call(session, "song.getData", api_token, {"SNG_ID": track_id})
+        sng_data = sng_data["results"]
+        track_token = sng_data["TRACK_TOKEN"]
+        url_data = {
+            "license_token": license_token,
+            "media": [
+                {
+                    "type": "FULL",
+                    "formats": [{"cipher": "BF_CBC_STRIPE", "format": "MP3_128"}],
```

Comment:
so, for our MVP release of deezer support the quality will be limited to 128kbit mp3 ?

Reply:
I will do my very best to solve this issue before. I started to observe the traffic of the App using the mitmproxy and I was able to get an old version of the deezer native plugin, which is a C-Library to build your own player. We will see, where this takes us 😅

Reply:
Oh and as a fallback solution we could also let the user store the arl-Cookie or his user data in the config. This way we could also get FLAC-Files

Reply:
Micha found a way to get Hi-Fi streaming to work. Really amazing!